### PR TITLE
fix:[CDS-85497]: replaced runningStages with current stage identifier in header

### PR DESCRIPTION
**Pull request by naman-goenka**
## Describe your changes
- **Bug** : In case of approved/rejected notification, in headers running stage set was being used. This created an issue when multiple running stages are present for example in looping, multi service, parallel stages, etc. Then incorrectly the header is like `Pipeline test current stage header  approved for stage: test current stage_0, test current stage_1, test current stage_2`. Also running stages will have additional metadata with enhanced notifications.
- **Expected Behaviour** :  The header should only have current stage name for which the approval notification is triggered.
- **Logic for current Stage Name** - Implemented by getting current stage level's identifier from ambiance. Hence searching for this identifier in layoutNodeMap results in getting name for that stage. 
- Testing: 
a Test multi service for running stage 
b Test looping flow for running stage 
c Test parallel flow for running stage 
d Test cd stage for running stage 
e Test approval stage for running stage 
f Test custom stage for running stage 
g Test when is include stage details setting is set to false by user
- Example notifications: 
[Looping](https://harness.atlassian.net/wiki/spaces/CDNG/pages/21563637966/Improved+harness+approval+notifications#:~:text=Looping%3A%20(Custom%20Stage)%20followed%20by%20looping%20approval%20stages%20followed%20by%20custom%20stage%20%5BFor%20checking%20header%20bug%20fix%5D)
[Parallel](https://harness.atlassian.net/wiki/spaces/CDNG/pages/21563637966/Improved+harness+approval+notifications#:~:text=Parallel%3A%20(Approval%20Stage)%20followed%20by%202%20parallel%20approval%20stages%20followed%20by%20custom%20stage%20%5BFor%20checking%20header%20bug%20fix%5D%20%3A%20check%20during%20parallel%20approval%20stage)

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56160)

### CODE REVIEW

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger codeformat, messagecheck, pmd

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger ut5, ut6, ut8, ut9

**Comment by krishi0408** [comment]
trigger codeformat

**Review by abhinavmittal-harness — APPROVED** [review]
> Reviewed 7 of 8 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56160)* status: 8 file(s) to review. 1 of 2 approvals obtained.

File: pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java [hunk 1]
```diff
@@ -207,6 +212,11 @@ public ApprovalSummary getApprovalSummary(Ambiance ambiance, HarnessApprovalInst
             .status(approvalInstance.getStatus())
             .pipelineExecutionLink(notificationHelper.generateUrl(ambiance))
             .timeRemainingForApproval(formatDuration(approvalInstance.getDeadline() - System.currentTimeMillis()))
+            .currentStageName(approvalInstance.isIncludePipelineExecutionHistory()
+                    ? getCurrentStageName(ambiance, pipelineExecutionSummaryEntity)
```
Comment (sahilhindwani-harness) [inline pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java:216]: where are we getting NPE in notification client. Should we fix there?

File: pipeline-service/service/src/main/resources/notification_templates/approval/email/plain_text_action_execution.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
-subject: "Pipeline ${pipelineName} ${status} for stage: ${runningStages}."
+subject: "Pipeline ${pipelineName} ${status} for stage: ${currentStageName}."
```
Comment (sahilhindwani-harness) [inline pipeline-service/service/src/main/resources/notification_templates/approval/email/plain_text_action_execution.txt:1]: Will this be a change in behaviour? Do we need to notify customer for the same?

**Review by sahilhindwani-harness — CHANGES_REQUESTED** [review]

File: pipeline-service/service/src/main/resources/notification_templates/approval/email/plain_text_action_execution.txt (hunk 2 above)
Comment (naman-goenka) [reply pipeline-service/service/src/main/resources/notification_templates/approval/email/plain_text_action_execution.txt:1]: This has been discussed and approved with product, it was a bug currently and was indicating false stages.

File: pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java (hunk 1 above)
Comment (naman-goenka) [reply pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java:216]: When trying to substitute with null values in template data. It expects not null values in template data.  
Should we raise a bug here?

Comment (naman-goenka) [reply pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java:216]: `io/harness/notification/channeldetails/EmailChannel.java:63`

proto's MapField$MutabilityAwareMap.putAll doesn't expects notNull field

Comment (sahilhindwani-harness) [reply pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java:216]: Yes, please raise a ticket for follow up.

**Review by sahilhindwani-harness — APPROVED** [review]

**Pull request merged**